gnomad/utils/file_utils.py

```diff
@@ -6,15 +6,93 @@
 import os
 import subprocess
 import uuid
-from typing import List, Optional, Tuple, Union
+from concurrent.futures import ThreadPoolExecutor
+from typing import Callable, List, Dict, Optional, Tuple, Union
 
 import hail as hl
+from hailtop.aiotools import LocalAsyncFS, RouterAsyncFS, AsyncFS
+from hailtop.aiogoogle import GoogleStorageAsyncFS
+from hailtop.utils import bounded_gather, tqdm
 
 logging.basicConfig(format="%(levelname)s (%(name)s %(lineno)s): %(message)s")
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
 
+async def parallel_file_exists(
+    fnames: List[str], parallelism: int = 750
+) -> Dict[str, bool]:
+    """
+    Check whether a large number of files exist.
+
+    Created for use with hail Batch jobs.
+    Normal `file_exists` function is very slow when checking a large number of files.
+
+    :param fnames: List of file names to check.
+    :param parallelism: Integer that sets parallelism of file existence checking task. Default is 750.
+    :return: Dictionary of file names (str) and whether the file exists (boolean).
+    """
+
+    async def low_level_async_file_exists(fs: AsyncFS, url: str) -> bool:
+        """
+        Check whether file exists.
+
+        :param fs: AsyncFS object.
+        :param url: File URL to check.
+        :return: Whether file URL exists.
+        """
+        try:
+            await fs.statfile(url)
+        except FileNotFoundError:
+            return False
+        else:
+            return True
+
+    async def async_file_exists(fs: AsyncFS, fname: str) -> bool:
+        """
+        Call `low_level_async_file_exists` to determine file existence.
+
+        :param fs: AsyncFS object.
+        :param fname: Path to file to check.
+        :return: Whether file exists.
+        """
+        fext = os.path.splitext(fname)[1]
+        if fext in [".ht", ".mt"]:
+            fname += "/_SUCCESS"
+        return await low_level_async_file_exists(fs, fname)
+
+    with tqdm(
+        total=len(fnames), desc="check files for existence", disable=False
+    ) as pbar:
+        with ThreadPoolExecutor() as thread_pool:
+            async with RouterAsyncFS(
+                "file", [LocalAsyncFS(thread_pool), GoogleStorageAsyncFS()]
+            ) as fs:
+
+                def create_unapplied_function(fname: str) -> Callable:
+                    """
+                    Create function to check if file exists and update progress bar in stdout.
+
+                    :param fname: Path to file to check.
+                    :return: Function that checks for file existence and updates progress bar.
+                    """
+
+                    async def unapplied_function():
+                        x = await async_file_exists(fs, fname)
+                        pbar.update(1)
+                        return x
+
+                    return unapplied_function
```

Comment:
A bit of context on this rigamarole:

When using asyncio, it's important to not have too many live coroutines. Every time you call an async function, you create a coroutine. For example, calling `create_unapplied_function('foo')` does not create a coroutine because we never invoke an async function. On the other hand, `create_unapplied_function('foo')()` *does* create a coroutine because it invokes the returned async function. A coroutine completes when we `await` it (or use one of the other asyncio waiting primitives, like `asyncio.wait`).

For this reason, `bounded_gather` accepts async functions as arguments instead of directly accepting coroutines. For example:
```python3
# this creates a coroutine for each argument to bad_bounded_gather
await bad_bounded_gather(async_do_something('argument'), ...)

# this allows bounded gather to bound the number of live coroutines (by waiting for
# previous coroutines to complete before invoking more)
async def async_delayed_do_something():
    return await async_do_something('argument')
await bounded_gather(async_delayed_do_something, ...)
```

In the first case, we create all the coroutines *before even calling `bad_bounded_gather`*. In the second case, `bounded_gather` can decide when to create coroutines (by invoking one of its arguments).

---

Usually, Hail team uses this pattern:

```python3
await bounded_gather(*[functools.partial(async_file_exists, fs, fname)
                       for fname in fnames],
                     parallelism=parallelism)
```

But we also want to update the progress bar after we check for file existence. I should maybe have called `create_unapplied_function` `check_existence_and_update_pbar_thunk`. I use the term "thunk" to refer to these zero-argument functions that delay the creation of coroutines.

Reply:
Thank you so much for this detailed description Dan!!

Reply:
seconded, thank you for this!! I added a small note to the docstring, but let me know if you think I should add a longer summary @jkgoodrich